### PR TITLE
Yuval/lbm1 4447 delete object store

### PR DESCRIPTION
**Pull request by yuval-lb**
I try to deal with this:
https://lightbitslabs.atlassian.net/browse/LBM1-4447

basically if we found more then 20 ERROR in osmosis output we raise Exception - 
catch that Exception and erase object store

then try again to do checkout

### CODE REVIEW

File: inaugurator/osmose.py [hunk 1]
```diff
@@ -49,9 +49,15 @@ def tellLabel(self, label):
 
     def wait(self):
         osmosis_output = []
+        error_counter = 0
+        TO_MANY_OBJECT_STORE_ERRORS = 20
```
Comment (shimon-lb) [inline inaugurator/osmose.py]: typo:
 TO => TOO

File: inaugurator/ceremony.py [hunk 2]
```diff
@@ -200,7 +200,10 @@ def _createBootAndInstallGrub(self, destination):
             with open(grubConfigFilename, "r") as grubConfigFile:
                 self._grubConfig = grubConfigFile.read()
 
-    def _osmosFromNetwork(self, destination):
+    def _osmosFromNetwork(self, destination, attempt=0):
+        if attempt > 1:
```
Comment (shimon-lb) [inline inaugurator/ceremony.py]: Please, define a macro, instead of hard-coded value.

File: inaugurator/osmose.py (hunk 1 above)
Comment (yuval-lb) [reply inaugurator/osmose.py]: +1

File: inaugurator/ceremony.py [hunk 3]
```diff
@@ -225,6 +228,10 @@ def _osmosFromNetwork(self, destination):
                 self._label = self._args.inauguratorNetworkLabel
             osmos.tellLabel(self._label)
             osmos.wait()
+        except osmose.CorruptedObjectStore:
+            logging.info("Found corrupted object store - Earsing everything!")
+            osmosiscleanup.OsmosisCleanup(destination, objectStorePath=self._localObjectStore).eraseEverything()
+            self._osmosFromNetwork(destination, attempt+1)
```
Comment (shimon-lb) [inline inaugurator/ceremony.py]: Consider splitting this logic and perform it by iterations, rather than a recursion...

File: inaugurator/osmose.py (hunk 1 above)
Comment (yuval-lb) [reply inaugurator/osmose.py]: osmosis might have other Errors he can handle, this specific error is throwing a lot of errors so 20 is a kind of "filter"

File: inaugurator/ceremony.py [hunk 4]
```diff
@@ -210,32 +210,39 @@ def _osmosFromNetwork(self, destination):
...
-            else:
-                try:
-                    self._talkToServer.failed(message=str(e))
-                except:
-                    pass
-            raise e
+        ATTEMPTS = 2
+        for attempt in range(ATTEMPTS):
+            try:
+                osmos = osmose.Osmose(
```
Comment (yuval-lb) [inline inaugurator/ceremony.py]: to help reviewers I will add:
nothing is changed here just was added in to a loop of 2 attempts.
and in general in this commit - we try first time, if we notice problems we clean the cache. then try again, if fail we clean again and raise.

File: inaugurator/ceremony.py [hunk 5]
```diff
@@ -210,32 +210,39 @@ def _osmosFromNetwork(self, destination):
...
+                    self._label = self._args.inauguratorNetworkLabel
+                osmos.tellLabel(self._label)
+                osmos.wait()
+            except osmose.CorruptedObjectStore:
+                    logging.info("Found corrupted object store - Earsing everything!")
+                    osmosiscleanup.OsmosisCleanup(destination, objectStorePath=self._localObjectStore).eraseEverything()
+                    if attempt > 0:
+                        raise Exception("Osmosis Excetpion! Could not checkout rootfs")
+            except Exception as e:
+                if self._debugPort is not None and self._debugPort.wasRebootCalled():
```
Comment (sashas-lb) [inline inaugurator/ceremony.py]: @yuval-lb those lines are not related to the commit .. please split this to another commit

Comment (sashas-lb) [reply inaugurator/ceremony.py]: ah .. now i see that you just relocated the code ..

**Comment by yuval-lb** [comment]
@sashas-lb update - 
its not working, I am not sure afte "eraseEverything" osmosis is still functional. seems like he gets an error and restart the machine.

I am having a lot of problems dubugging this using Personalab.

**Comment by sashas-lb** [comment]
Hi sorry just saw it now ..I can try to help tomorrow

**Comment by yuval-lb** [comment]
Dear reviewers @sashas-lb @shimon-lb the code have been update and was changed, it is not very long please re review the PR.

its already tested and proved to be working.

File: inaugurator/osmosiscleanup.py [hunk 6]
```diff
@@ -35,4 +35,12 @@ def _attemptObjectStoreCleanup(self):
             pass
 
     def eraseEverything(self):
-        sh.run("busybox rm -fr %s/*" % self._objectStore.root())
+        try:
+            sh.run("busybox rm -fr %s/*" % self._objectStore.root())
+        except Exception as e:
+            try:
+                sh.run("busybox rm -fr %s/*" % self._objectStore.root())  # because - http://lists.busybox.net/pipermail/busybox/2015-August/083233.html
```
[the remark this answers is not in the document]
Comment (yuval-lb) [reply inaugurator/osmosiscleanup.py]: running twice simply works. as it shown in the open bug -http://lists.busybox.net/pipermail/busybox/2015-August/083233.html. 
when its happening we its printing to the screen the info, which files he didnt manged to delete.
and sh.run also print the command its running.

the fs is rw, we are doing alot of stuff before we arrive to this stage

File: inaugurator/osmose.py [hunk 7]
```diff
@@ -47,13 +47,21 @@ def tellLabel(self, label):
         self._popen.stdin.write(label + "\n")
         self._popen.stdin.close()
 
-    def wait(self):
+    def wait(self, inspect_erros=False):
         osmosis_output = []
+        error_counter = 0
+        TOO_MANY_OBJECT_STORE_ERRORS = 20
```
[the remark this answers is not in the document]
Comment (yuval-lb) [reply inaugurator/osmose.py]: in general whats happening, our "bug" is actually a feature of osmosis - he can identify bad hash and fix them.
the problem is that he thinks that all hashs are corrupted then this process is taking alot of time.

if for some reason we have small chunk of of bad hashs which osmosis can fix, I dont want to interupt it.
but still 20 is full random.

anyway if osmosis will exit with return code !=0 we will delete the cache

**Comment by yuval-lb** [comment]
guys, I am guessing this issue is wasting a lot of time to developers, lets try to progress with this
@maor-lb @ira-lb @sashas-lb @shimon-lb

File: inaugurator/osmosiscleanup.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (sashas-lb) [reply inaugurator/osmosiscleanup.py]: @yuval-lb  pleaes put a fat commect on why are you doing it .. it is very .. unexpected

File: inaugurator/osmose.py [hunk 8]
```diff
@@ -47,13 +47,21 @@ def tellLabel(self, label):
         self._popen.stdin.write(label + "\n")
```
Comment (sashas-lb) [inline inaugurator/osmose.py]: LGTM

File: inaugurator/ceremony.py (hunk 4 above)
Comment (sashas-lb) [reply inaugurator/ceremony.py]: can you please split it like this : 
1) first commit extracts osmosis stuff into separate method 
2) second commit does the loop .. 

it is hard to review as it is not clear what is been actually changed

File: inaugurator/ceremony.py [hunk 9]
```diff
@@ -210,32 +210,44 @@ def _osmosFromNetwork(self, destination):
...
+                if attempt == 0:
+                    osmos.wait(inspect_erros=True)
+                else:
+                    osmos.wait()
+                return
+            except osmose.CorruptedObjectStore:
+                logging.info("Found corrupted object store - purge osmosis!")
                 try:
-                    self._talkToServer.failed(message=str(e))
+                    osmosiscleanup.OsmosisCleanup(destination, objectStorePath=os.path.join(destination,"var", "lib", "osmosis", "objectstore")).eraseEverything()
```
Comment (sashas-lb) [inline inaugurator/ceremony.py]: please extract path to variable before using it

File: inaugurator/ceremony.py (hunk 4 above)
Comment (sashas-lb) [reply inaugurator/ceremony.py]: @yuval-lb i also find it hard to understand hopw exception will be raised in second time ..

**Comment by sashas-lb** [comment]
Cool .. ping me when done I will take another look

**Comment by yuval-lb** [comment]
@sashas-lb ping

File: inaugurator/ceremony.py [hunk 10]
```diff
@@ -210,6 +210,10 @@ def _osmosFromNetwork(self, destination):
             self._talkToServer = talktoserver.TalkToServer(
                 amqpURL=self._args.inauguratorServerAMQPURL, myID=self._args.inauguratorMyIDForServer)
             self._talkToServer.checkIn()
+            message = self._talkToServer.label()
+            self._label = json.loads(message)['rootfs']
+        else:
```
Comment (sashas-lb) [inline inaugurator/ceremony.py]: how do we get into else ? why would 
"self._args.inauguratorServerAMQPURL" be None?

Comment (sashas-lb) [reply inaugurator/ceremony.py]: ah .. sorry .. now i see that this is not a logics change

File: inaugurator/ceremony.py [hunk 11]
```diff
@@ -214,27 +214,56 @@ def _osmosFromNetwork(self, destination):
...
-            if self._debugPort is not None and self._debugPort.wasRebootCalled():
-                logging.info("Waiting to be reboot (from outside)...")
-                blockForever = threading.Event()
-                blockForever.wait()
-            else:
+        ATTEMPTS = 2
+        for attempt in range(ATTEMPTS):
+            try:
+                if attempt == 0:
+                    self._checkoutOsmosFromNetwork(destination,
```
Comment (sashas-lb) [inline inaugurator/ceremony.py]: if we have this "if" why dont we extract this outside "for" loop ?

**Comment by sashas-lb** [comment]
looks much better .. there are some small comments .. to your discretion ..

**Comment by yuval-lb** [comment]
this is a tested, working version.
latest upadate:
function:
self._checkoutOsmosFromNetwork
second run called with:
talkeToServer=None

as it happend in the first run, running reportToServer again raise exception.

**Comment by yuval-lb** [comment]
@ira-lb can you remove your "changes requested"

File: inaugurator/osmosiscleanup.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ira-lb) [reply inaugurator/osmosiscleanup.py]: Can we have that Comment now? I want code we can maintain and pass on to newbies. Otherwise call it yet_another_hackity_hack.

File: inaugurator/osmose.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ira-lb) [reply inaugurator/osmose.py]: OK, so please add this remark in the code.

File: inaugurator/osmosiscleanup.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (yuval-lb) [reply inaugurator/osmosiscleanup.py]: I put the link to the bug in busybox, whats more I can add ?

**Comment by abel-lb** [comment]
@yogev-lb is it possible to report this specific error somehow to the lab stats @yogev-lb pulls and publish in graphite dashboard ?

File: inaugurator/osmose.py [hunk 12]
```diff
@@ -51,6 +51,10 @@ def wait(self, inspect_erros=False):
         osmosis_output = []
         error_counter = 0
         TOO_MANY_OBJECT_STORE_ERRORS = 20
+        '''
+        osmosis - can identify bad hash and fix them.
+        the problem is that if he thinks that all the hashs are corrupted then this process will take a lot of time.
+        '''
```
Comment (ira-lb) [inline inaugurator/osmose.py]: but why 20 and not 34½?

Comment (yuval-lb) [reply inaugurator/osmose.py]: as I wrote earlier its random, do you have a better suggestion?

Comment (ira-lb) [reply inaugurator/osmose.py]: yes, a short explanation why 20 was picked. why not 1 or 100. how do we know it's an optimal value?
